pennylane/utils.py

```diff
@@ -433,3 +434,77 @@ def expand_vector(vector, original_wires, expanded_wires):
     expanded_tensor = np.moveaxis(expanded_tensor, original_indices, wire_indices)
 
     return expanded_tensor.reshape(2 ** M)
+
+
+class Spy:
+    """Context in which calls to a QubitDevice's execute method are counted without performing the execution.
+
+    The class mocks the execute method, returning a dummy value of the correct dimensions.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.utils import Spy
+        qml.enable_tape()
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with Spy() as spy:
+            circuit(0.1)
+            circuit(0.5)
+
+        print(spy.counts) # 2
+
+        with Spy() as spy_grad:
+            g = qml.grad(circuit)
+            g(0.1)
+
+        # parameter-shift rule uses 2+1 executions
+        # to compute the gradient
+        print(spy_grad.counts) # 3
+
+    """
+
+    def __init__(self):
+        self.counts = None
+        self.temp_original = None
+
+    def __enter__(self):
+        # remember the original method
+        self.temp_original = qml._qubit_device.QubitDevice.execute
+
+        # Mock qubit device's execute method.
+        # The side effect will define the output of the mock.
+        qml._qubit_device.QubitDevice.execute = Mock(side_effect=self.side_effect)
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+
+        # record the number of calls to the mocked function
+        self.counts = qml._qubit_device.QubitDevice.execute.call_count
+
+        # restore the original function
+        qml._qubit_device.QubitDevice.execute = self.temp_original
+
+    @staticmethod
+    def side_effect(circuit, **kwargs):  # pylint: disable=unused-argument
+        """Determine mock output. This function accepts the same inputs as the mocked function.
+
+        Args:
+            circuit (.tape.QuantumTape): tape representing the circuit
+
+        Returns:
+            np.array: dummy output
+        """
+
+        dim_output = len(circuit.measurements)
+        return np.zeros(shape=(dim_output,))
```

Comment:
Perhaps better to use `tape.output_dim` here? This takes into account things like probabilities and states, where the number of dimensions is no longer equal to the number of measurements.

Reply:
Also, you might have to test if it works with samples 😬